spacy/pipeline/entity_linker.py

```diff
@@ -220,11 +218,13 @@ def update(
             return losses
         validate_examples(examples, "EntityLinker.update")
         sentence_docs = []
-        docs = [eg.predicted for eg in examples]
-        if set_annotations:
-            # This seems simpler than other ways to get that exact output -- but
-            # it does run the model twice :(
-            predictions = self.model.predict(docs)
+        docs = []
+        for eg in examples:
+            eg.predicted.ents = eg.reference.ents
+            docs.append(eg.predicted)
+        # This seems simpler than other ways to get that exact output -- but
+        # it does run the model twice :(
+        predictions = self.predict(docs)
```

Comment:
Hmm, I see why this ends up convenient based on the code in the `predict()` method, but I hope we can refactor that in a separate PR at some point. I might have a look today.

We want to get to a point where we're doing something like this:

```python

sentence_encodings, bp_context = self.model.begin_update(sentence_docs)
predictions = self._encodings2predictions(sentence_encodings, sentence_docs)
self.set_annotations(predictions)
```

This would also get us to where we're calling the model on the full batch of data, instead of per-sentence as we are currently.

Reply:
Yea I know, I was looking at this code when making the PR and thinking this should really be cleaned up further :| (but wanted this PR to go out first)